### ISSUE TITLE
[Time Picker / Input Time Picker / Input type="time"] Handle 24 hour format

### ISSUE THREAD

**Issue by macandcheese** [Esri/calcite-design-system#4756]
### Description

Currently, the time-picker, input-time-picker, and input type="time" do not allow for the use of 24 hour time format from an end-user use case.

### Acceptance Criteria

Allow the ability to control the format - 12/24hr, in these components / component configurations.

### Relevant Info

A native (non-calcite) input of type="time" automatically changes from 12 -> 24 hour format when I changed my Time Format in OS preferences to 24-hour time (Mac / Chrome). I could not replicate this behavior with calcite-input of type time.

### Which Component

[Time Picker / Input Time Picker / Input type="time"]

**Comment by jcfranco** [comment]
@annierm18 Can you confirm if this needs to be configurable? So far, the format has been automatically determined by the locale.

**Comment by eriklharper** [comment]
Sounds like the issue is just that the component doesn't respond to browser or OS locale changes properly.  Is that correct?

**Comment by eriklharper** [comment]
From the looks of it, this doesn't seem possible to do with the existing browser APIs.  I'm going to do some more research, but I don't think there's a way to access this system setting via javascript in the browser.

**Comment by benelan** [comment]
IIRC `navigator.language` should be the system's preferred language when provided.

Last time I looked into this I found it really depends on the browser and OS. For example, FireFox:

> All major operating systems have a Settings UI for selecting those preferences, and since Firefox does not provide its own, Gecko looks into the OS for them. A special API `mozilla::intl::OSPreferences` handles communication with the host operating system, retrieving regional preferences and altering internationalization formatting with user preferences.

 https://firefox-source-docs.mozilla.org/intl/locale.html#regional-preferences

@macandcheese said he is using Mac/Chrome:
https://developer.chrome.com/docs/extensions/reference/i18n/#mac-os-x

I know on Windows you can turn the option on/off thru the system settings
![image](https://user-images.githubusercontent.com/10986395/178860506-e2de7e94-0bce-4e47-8fe3-2e1f796a8927.png)

I don't think input uses `navigator.language`

**Comment by eriklharper** [comment]
Yeah, and even if the specific browser provides a hook into the OS locale, which would imply a default hour cycle, it would still need to hook into the OS hour cycle checkbox to determine if a user has overridden the default.

**Comment by annierm18** [comment]
Thank you for the info @eriklharper and @benelan! I will discuss this with Babak and will follow up soon.

**Comment by annierm18** [comment]
Update on this - the i18n team would like for this to be implemented if it's possible.

**Comment by eriklharper** [comment]
> Update on this - the i18n team would like for this to be implemented if it's possible.

What we can support is an `hour-cycle` property, but we won't be able to set the hour cycle based on the user's OS setting.  I can look into doing this in browser-specific time format settings, but after the research I did I discovered the browsers don't have a hook into the OS-level system preferences like I assumed they would.  Perhaps in the future this is something they will implement.

Would a property that overrides the existing locale's hour cycle suffice?

**Comment by benelan** [comment]
> What we can support is an hour-cycle property, but we won't be able to set the hour cycle based on the user's OS setting.

I'm pretty sure it's possible using `navigator.language` like I mentioned above. What about something like [this](https://stackoverflow.com/a/64480119)?

```js
/*
 * Detects navigator locale 24h time preference
 * It works by checking whether hour output contains AM ('1 AM' or '01 h')
 * based on the user's preferred language
 */
const isBrowserLocale24h = () =>
  !new Intl.DateTimeFormat(navigator.language, { hour: "numeric" })
    .format(0)
    .match(/AM/);
```

**Comment by benelan** [comment]
Looking at the guys code you would probably have to check for "PM" too:
```
const isBrowserLocale24h = () =>
  !new Intl.DateTimeFormat(navigator.language, { hour: "numeric" })
    .format(0)
    .match(/[AM|PM]/);

**Comment by eriklharper** [comment]
Correct, I wouldn't expect this to work anyway because `Intl.DateTimeFormat` is just a formatting api.  It knows nothing about user preferences.

**Comment by eriklharper** [comment]
That code snippet won't work consistently enough to rely on it.  Didn't work for me when I changed my OS time preference to 24 hour with an en-US locale.

**Comment by eriklharper** [comment]
The kicker there is "_default_ system language".  The trick is getting it to recognize the non-default hour setting, which it doesn't appear to support.

**Comment by driskull** [comment]
Can we use this? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/hourCycle

**Comment by eriklharper** [comment]
> Can we use this? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/hourCycle

Yes, we would use this to set the default value of the `hour-cycle` prop based on the locale.  Being a prop, `hour-cycle` would also be overridable from the locale's default hour cycle.

**Comment by geospatialem** [comment]
Reallocated to October to accommodate a higher request moderate priority from the Maps SDK for JS team.

**Comment by geospatialem** [comment]
Updated the milestone to December, and upgraded the priority to medium (previously low) to ensure this enhancement lands prior to the 24R1 release. cc @hgonzago and @JonUihlein

**Comment by eriklharper** [comment]
Update:  Almost done writing tests for all the supported locales, and looking at a handful of some locale-specific issues with typing in localized meridiem characters and one with the arab numbering-system, and then refactoring from `hour-cycle` to `hour-format` before this is ready.  Getting close :)

**Comment by geospatialem** [comment]
@eriklharper I'm encountering what _may_ be unexpected behavior with the component when `hour-format="24"` and entering a number over 12 for the hour. Once the minute is added the value appears accurate, but prior to selection it doesn't match the number, nor are numbers 13+ added to the component initially. Is this expected?

[Sample with `3.0.0-next.87`](https://codepen.io/geospatialem/pen/RNbjWoM) and a screen grab of the behavior 
![verify-4756](https://github.com/user-attachments/assets/37d5a143-48ab-4544-b918-b3c603a1584d)

**Comment by eriklharper** [comment]
> @eriklharper I'm encountering what _may_ be unexpected behavior with the component when `hour-format="24"` and entering a number over 12 for the hour. Once the minute is added the value appears accurate, but prior to selection it doesn't match the number, nor are numbers 13+ added to the component initially. Is this expected?
> 
> [Sample with `3.0.0-next.87`](https://codepen.io/geospatialem/pen/RNbjWoM) and a screen grab of the behavior ![verify-4756](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4NTg0ODUsIm5iZiI6MTczNTg1ODE4NSwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDJUMjI0OTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDI3Njg2ZDBhMTNjNTE3Yjk0YzljMTlmNTg0ZTMzNDcwNWVjOWNiODFkOTFmODFjNjBkY2I4Nzg1ODNmOTkxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Qqd7tYQV0qGcfg9tl-VFuJV7Wv7ePOnpFw2zPlEReIE) [ ![verify-4756](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4NTg0ODUsIm5iZiI6MTczNTg1ODE4NSwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDJUMjI0OTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDI3Njg2ZDBhMTNjNTE3Yjk0YzljMTlmNTg0ZTMzNDcwNWVjOWNiODFkOTFmODFjNjBkY2I4Nzg1ODNmOTkxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Qqd7tYQV0qGcfg9tl-VFuJV7Wv7ePOnpFw2zPlEReIE) ](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4NTg0ODUsIm5iZiI6MTczNTg1ODE4NSwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDJUMjI0OTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDI3Njg2ZDBhMTNjNTE3Yjk0YzljMTlmNTg0ZTMzNDcwNWVjOWNiODFkOTFmODFjNjBkY2I4Nzg1ODNmOTkxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Qqd7tYQV0qGcfg9tl-VFuJV7Wv7ePOnpFw2zPlEReIE) [ ](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4NTg0ODUsIm5iZiI6MTczNTg1ODE4NSwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDJUMjI0OTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDI3Njg2ZDBhMTNjNTE3Yjk0YzljMTlmNTg0ZTMzNDcwNWVjOWNiODFkOTFmODFjNjBkY2I4Nzg1ODNmOTkxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Qqd7tYQV0qGcfg9tl-VFuJV7Wv7ePOnpFw2zPlEReIE)

Nice catch there.  Not expected for sure.  Seems to be an initial state kind of behavior where the formatting is displaying as the default 12-hour even though the correct 24-hour hour value is being correctly stored.  Looks like that behavior ceases in subsequent updates, but agreed that this isn't expected.  I would call this a bug for sure, probably low-ish on the severity, but definitely something we should fix.

**Comment by geospatialem** [comment]
> @eriklharper I'm encountering what _may_ be unexpected behavior with the component when `hour-format="24"` and entering a number over 12 for the hour. Once the minute is added the value appears accurate, but prior to selection it doesn't match the number, nor are numbers 13+ added to the component initially. Is this expected?
> 
> [Sample with `3.0.0-next.87`](https://codepen.io/geospatialem/pen/RNbjWoM) and a screen grab of the behavior ![verify-4756](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU5MTcyMzQsIm5iZiI6MTczNTkxNjkzNCwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDNUMTUwODU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmIxZTg2YjU0Y2Q1YzJjMTUwY2E2MWIxMWY3MjIzNWFhNTA1MTIxOWQzZmI1YjE3MTc0Yzg5ZmE3N2U2MWZiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FCHoWXdBtFvbZHUorP3gHYhytc0-Ugtxg90mDzjHt8g) [ ![verify-4756](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU5MTcyMzQsIm5iZiI6MTczNTkxNjkzNCwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDNUMTUwODU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmIxZTg2YjU0Y2Q1YzJjMTUwY2E2MWIxMWY3MjIzNWFhNTA1MTIxOWQzZmI1YjE3MTc0Yzg5ZmE3N2U2MWZiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FCHoWXdBtFvbZHUorP3gHYhytc0-Ugtxg90mDzjHt8g) ](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU5MTcyMzQsIm5iZiI6MTczNTkxNjkzNCwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDNUMTUwODU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmIxZTg2YjU0Y2Q1YzJjMTUwY2E2MWIxMWY3MjIzNWFhNTA1MTIxOWQzZmI1YjE3MTc0Yzg5ZmE3N2U2MWZiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FCHoWXdBtFvbZHUorP3gHYhytc0-Ugtxg90mDzjHt8g) [ ](https://private-user-images.githubusercontent.com/5023024/399804403-37d5a143-48ab-4544-b918-b3c603a1584d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU5MTcyMzQsIm5iZiI6MTczNTkxNjkzNCwicGF0aCI6Ii81MDIzMDI0LzM5OTgwNDQwMy0zN2Q1YTE0My00OGFiLTQ1NDQtYjkxOC1iM2M2MDNhMTU4NGQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDNUMTUwODU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmIxZTg2YjU0Y2Q1YzJjMTUwY2E2MWIxMWY3MjIzNWFhNTA1MTIxOWQzZmI1YjE3MTc0Yzg5ZmE3N2U2MWZiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FCHoWXdBtFvbZHUorP3gHYhytc0-Ugtxg90mDzjHt8g)

Created #11198 to address the unrelated behavior for 25.R1. Closing out the remaining effort above as complete and verified. 🎉
